File: packages/engine/src/standings.ts
import { modeInfo } from './modes';
import type { GameState, LeaderboardMode, Match, Player, Round, Standing } from './types';

export function isScored(m: Match): boolean {
  return m.scoreA !== null && m.scoreB !== null;
}

export function isRoundComplete(r: Round): boolean {
  return r.matches.every(isScored);
}

/** 1 = team A won, -1 = team B won, 0 = draw. */
export function matchOutcome(m: Match): 1 | -1 | 0 {
  const a = m.scoreA ?? 0;
  const b = m.scoreB ?? 0;
  return a > b ? 1 : a < b ? -1 : 0;
}

interface Acc {
  id: string;
  name: string;
  playerIds: string[];
  played: number;
  wins: number;
  draws: number;
  losses: number;
  pointsFor: number;
  pointsAgainst: number;
  byes: number;
}

function unitKey(state: GameState, playerId: string, byPlayer: Map<string, Player>): string {
  return modeInfo(state.settings.mode).teams ? byPlayer.get(playerId)!.teamId! : playerId;
}

export function teamName(players: Player[], ids: string[]): string {
  return ids.map((id) => players.find((p) => p.id === id)?.name ?? '?').join(' & ');
}

/**
 * Standings from every scored match in rounds with index < `uptoRound`
 * (all rounds by default). Byes count only for completed rounds.
 */
export function computeStandings(state: GameState, uptoRound = Infinity): Standing[] {
  const current = rank(state, accumulate(state, uptoRound));
  const completed = state.rounds.filter((r) => r.index < uptoRound && isRoundComplete(r));
  if (completed.length < 2) return current;

  const lastDone = completed[completed.length - 1].index;
  const previous = rank(state, accumulate(state, lastDone));
  const prevRank = new Map(previous.map((s) => [s.id, s.rank]));
  return current.map((s) => ({ ...s, movement: (prevRank.get(s.id) ?? s.rank) - s.rank }));
}

function accumulate(state: GameState, uptoRound: number): Map<string, Acc> {
  const byPlayer = new Map(state.players.map((p) => [p.id, p]));
  const acc = new Map<string, Acc>();
  const teams = modeInfo(state.settings.mode).teams;

  for (const p of state.players) {
    const key = unitKey(state, p.id, byPlayer);
    const entry = acc.get(key);
    if (entry) {
      entry.playerIds.push(p.id);
      continue;
    }
    acc.set(key, {
      id: key,
      name: p.name,
      playerIds: [p.id],
      played: 0,
      wins: 0,
      draws: 0,
      losses: 0,
      pointsFor: 0,
      pointsAgainst: 0,
      byes: 0,
    });
  }
  if (teams) for (const a of acc.values()) a.name = teamName(state.players, a.playerIds);

  for (const round of state.rounds) {
    if (round.index >= uptoRound) continue;
    for (const m of round.matches) {
      if (!isScored(m)) continue;
      const outcome = matchOutcome(m);
      credit(acc, teams ? [unitKey(state, m.teamA[0], byPlayer)] : m.teamA, m.scoreA!, m.scoreB!, outcome);
      credit(acc, teams ? [unitKey(state, m.teamB[0], byPlayer)] : m.teamB, m.scoreB!, m.scoreA!, -outcome as 1 | -1 | 0);
    }
    if (isRoundComplete(round)) {
      const byeUnits = new Set(round.byes.map((id) => unitKey(state, id, byPlayer)));
      for (const u of byeUnits) acc.get(u)!.byes += 1;
    }
  }
  return acc;
}

function credit(acc: Map<string, Acc>, ids: string[], pf: number, pa: number, outcome: 1 | -1 | 0) {
  for (const id of ids) {
    const a = acc.get(id)!;
    a.played += 1;
    a.pointsFor += pf;
    a.pointsAgainst += pa;
    if (outcome === 1) a.wins += 1;
    else if (outcome === -1) a.losses += 1;
    else a.draws += 1;
  }
}

function rank(state: GameState, acc: Map<string, Acc>): Standing[] {
  const { settings } = state;
  const mode: LeaderboardMode = settings.scoring.type === 'off' ? 'wins' : settings.leaderboard;
  const fallbackAvg = (settings.scoring.points ?? 0) / 2;

  const rows = [...acc.values()].map((a) => {
    const avg = a.played ? a.pointsFor / a.played : 0;
    const byeComp = settings.byePoints === 'average' ? a.byes * (a.played ? avg : fallbackAvg) : 0;
    const total = a.pointsFor + byeComp;
    const winScore = a.wins + a.draws / 2;
    const score =
      mode === 'points' ? round2(total) : mode === 'wins' ? winScore : round2(a.played ? total / (a.played + (byeComp ? a.byes : 0)) : 0);
    const diff = a.pointsFor - a.pointsAgainst;
    const keys =
      mode === 'points'
        ? [score, diff, a.wins]
        : mode === 'wins'
          ? [winScore, diff, a.pointsFor]
          : [score, a.played ? diff / a.played : 0, winScore];
    return { a, score, diff, keys };
  });

  rows.sort((x, y) => {
    for (let i = 0; i < x.keys.length; i++) {
      if (x.keys[i] !== y.keys[i]) return y.keys[i] - x.keys[i];
    }
    return x.a.name.localeCompare(y.a.name);
  });

  let lastKeys: number[] | null = null;
  let lastRank = 0;
  return rows.map((r, i) => {
    const same = lastKeys && lastKeys.every((k, j) => k === r.keys[j]);
    const rankValue = same ? lastRank : i + 1;
    lastKeys = r.keys;
    lastRank = rankValue;
    const { a } = r;
    return {
      id: a.id,
      name: a.name,
      playerIds: a.playerIds,
      rank: rankValue,
      played: a.played,
      wins: a.wins,
      draws: a.draws,
      losses: a.losses,
      pointsFor: a.pointsFor,
      pointsAgainst: a.pointsAgainst,
      diff: r.diff,
      byes: a.byes,
      score: r.score,
      movement: 0,
    };
  });
}

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}

File: apps/mobile-shared/src/commonMain/kotlin/app/padel/engine/Standings.kt
package app.padel.engine

import kotlin.math.floor

fun isScored(m: Match): Boolean = m.scoreA != null && m.scoreB != null

fun isRoundComplete(r: Round): Boolean = r.matches.all(::isScored)

/** 1 = team A won, -1 = team B won, 0 = draw. */
fun matchOutcome(m: Match): Int {
    val a = m.scoreA ?: 0
    val b = m.scoreB ?: 0
    return if (a > b) 1 else if (a < b) -1 else 0
}

private class Acc(
    val id: String,
    var name: String,
    val playerIds: MutableList<String>,
) {
    var played = 0
    var wins = 0
    var draws = 0
    var losses = 0
    var pointsFor = 0
    var pointsAgainst = 0
    var byes = 0
}

private fun unitKey(state: GameState, playerId: String, byPlayer: Map<String, Player>): String =
    if (modeInfo(state.settings.mode).teams) byPlayer.getValue(playerId).teamId!! else playerId

fun teamName(players: List<Player>, ids: List<String>): String =
    ids.joinToString(" & ") { id -> players.find { it.id == id }?.name ?: "?" }

/**
 * Standings from every scored match in rounds with index < [uptoRound]
 * (all rounds by default). Byes count only for completed rounds.
 */
fun computeStandings(state: GameState, uptoRound: Int = Int.MAX_VALUE): List<Standing> {
    val current = rank(state, accumulate(state, uptoRound))
    val completed = state.rounds.filter { it.index < uptoRound && isRoundComplete(it) }
    if (completed.size < 2) return current

    val lastDone = completed.last().index
    val previous = rank(state, accumulate(state, lastDone))
    val prevRank = previous.associate { it.id to it.rank }
    return current.map { it.copy(movement = (prevRank[it.id] ?: it.rank) - it.rank) }
}

private fun accumulate(state: GameState, uptoRound: Int): Map<String, Acc> {
    val byPlayer = state.players.associateBy { it.id }
    val acc = linkedMapOf<String, Acc>()
    val teams = modeInfo(state.settings.mode).teams

    for (p in state.players) {
        val key = unitKey(state, p.id, byPlayer)
        val entry = acc[key]
        if (entry != null) {
            entry.playerIds += p.id
            continue
        }
        acc[key] = Acc(key, p.name, mutableListOf(p.id))
    }
    if (teams) for (a in acc.values) a.name = teamName(state.players, a.playerIds)

    for (round in state.rounds) {
        if (round.index >= uptoRound) continue
        for (m in round.matches) {
            if (!isScored(m)) continue
            val outcome = matchOutcome(m)
            credit(acc, if (teams) listOf(unitKey(state, m.teamA[0], byPlayer)) else m.teamA, m.scoreA!!, m.scoreB!!, outcome)
            credit(acc, if (teams) listOf(unitKey(state, m.teamB[0], byPlayer)) else m.teamB, m.scoreB, m.scoreA, -outcome)
        }
        if (isRoundComplete(round)) {
            val byeUnits = round.byes.map { unitKey(state, it, byPlayer) }.toSet()
            for (u in byeUnits) acc.getValue(u).byes += 1
        }
    }
    return acc
}

private fun credit(acc: Map<String, Acc>, ids: List<String>, pf: Int, pa: Int, outcome: Int) {
    for (id in ids) {
        val a = acc.getValue(id)
        a.played += 1
        a.pointsFor += pf
        a.pointsAgainst += pa
        when (outcome) {
            1 -> a.wins += 1
            -1 -> a.losses += 1
            else -> a.draws += 1
        }
    }
}

private class Row(val a: Acc, val score: Double, val diff: Int, val keys: List<Double>)

private fun rank(state: GameState, acc: Map<String, Acc>): List<Standing> {
    val settings = state.settings
    val mode = if (settings.scoring.type == ScoringType.Off) LeaderboardMode.Wins else settings.leaderboard
    val fallbackAvg = (settings.scoring.points ?: 0) / 2.0

    val rows = acc.values.map { a ->
        val avg = if (a.played > 0) a.pointsFor.toDouble() / a.played else 0.0
        val byeComp = if (settings.byePoints == ByePoints.Average) a.byes * (if (a.played > 0) avg else fallbackAvg) else 0.0
        val total = a.pointsFor + byeComp
        val winScore = a.wins + a.draws / 2.0
        val score = when (mode) {
            LeaderboardMode.Points -> round2(total)
            LeaderboardMode.Wins -> winScore
            LeaderboardMode.Average ->
                round2(if (a.played > 0) total / (a.played + (if (byeComp != 0.0) a.byes else 0)) else 0.0)
        }
        val diff = a.pointsFor - a.pointsAgainst
        val keys = when (mode) {
            LeaderboardMode.Points -> listOf(score, diff.toDouble(), a.wins.toDouble())
            LeaderboardMode.Wins -> listOf(winScore, diff.toDouble(), a.pointsFor.toDouble())
            LeaderboardMode.Average -> listOf(score, if (a.played > 0) diff.toDouble() / a.played else 0.0, winScore)
        }
        Row(a, score, diff, keys)
    }

    val sorted = rows.sortedWith { x, y ->
        for (i in x.keys.indices) {
            if (x.keys[i] != y.keys[i]) return@sortedWith y.keys[i].compareTo(x.keys[i])
        }
        localeCompare(x.a.name, y.a.name)
    }

    var lastKeys: List<Double>? = null
    var lastRank = 0
    return sorted.mapIndexed { i, r ->
        val same = lastKeys == r.keys
        val rankValue = if (same) lastRank else i + 1
        lastKeys = r.keys
        lastRank = rankValue
        val a = r.a
        Standing(
            id = a.id,
            name = a.name,
            playerIds = a.playerIds.toList(),
            rank = rankValue,
            played = a.played,
            wins = a.wins,
            draws = a.draws,
            losses = a.losses,
            pointsFor = a.pointsFor,
            pointsAgainst = a.pointsAgainst,
            diff = r.diff,
            byes = a.byes,
            score = r.score,
            movement = 0,
        )
    }
}

/** `Math.round(n * 100) / 100` — JS rounds halves towards +∞. */
private fun round2(n: Double): Double = floor(n * 100 + 0.5) / 100

/**
 * Approximates `String.prototype.localeCompare` (ICU root collation) for player names:
 * punctuation and spaces sort before digits and letters, letters compare case-insensitively
 * first, and lower case sorts before upper case on an otherwise equal name.
 */
internal fun localeCompare(a: String, b: String): Int {
    fun weight(c: Char): Int = when {
        c.isLetter() -> 2
        c.isDigit() -> 1
        else -> 0
    }
    val n = minOf(a.length, b.length)
    for (i in 0 until n) {
        val x = a[i]
        val y = b[i]
        val wx = weight(x)
        val wy = weight(y)
        if (wx != wy) return wx.compareTo(wy)
        val c = x.lowercaseChar().compareTo(y.lowercaseChar())
        if (c != 0) return c.coerceIn(-1, 1)
    }
    if (a.length != b.length) return a.length.compareTo(b.length)
    for (i in 0 until n) {
        if (a[i] != b[i]) return if (a[i].isLowerCase()) -1 else 1
    }
    return 0
}
